File: lintcode/python/0152_combinations.py
#!/usr/bin/python -t

class Solution:
    """
    @param n: Given the range of numbers
    @param k: Given the numbers of combinations
    @return: All the combinations of k numbers out of 1..n
    """
    def combine(self, n, k):
        # write your code here
        self.ret = []
        self.dfs(n, k, 1, 0, [])
        
        return self.ret
        
    def dfs(self, n, k, index, cnt, path):
        if cnt == k:
            self.ret.append(list(path))
            return
        
        for i in range(index, n+1):
            if i not in path:
                path.append(i)
                self.dfs(n, k, i+1, cnt+1, path)
                path.pop()
                
        return
            

File: lintcode/golang/0152_combinations.go
/**
 * @param n: Given the range of numbers
 * @param k: Given the numbers of combinations
 * @return: All the combinations of k numbers out of 1..n
 */
func combine (n int, k int) [][]int {
    // write your code here
    if n == 0 {
        return [][]int{}
    }
    
    ret := [][]int{}
    
    var dfs func(int, int, int, []int, *[][]int)
    dfs = func (n int, k int, index int, path []int, ret *[][]int) {
        if k == len(path) {
            tmp := make([]int, len(path))
            copy(tmp, path)
            *ret = append(*ret, tmp)
            return
        }
        
        for i := index; i <= n; i++ {
            path = append(path, i)
            dfs(n, k, i+1, path, ret)
            path = path[:len(path)-1]
        }
    }
    
    dfs(n, k, 1, []int{}, &ret)
    
    return ret
}

